### ISSUE TITLE
Voice recognition

### ISSUE THREAD

**Issue by TheBest-Name** [OHF-Voice/linux-voice-assistant#240]
I saw this in HA discussions:
https://github.com/orgs/home-assistant/discussions/527#discussion-8675242

I didn't make it, but I saw it and I think it would be a great addition to this. I feel like it would also be implemented here as well.

I am mainly putting it here also because I don't care for the discussions for feature requests in HA. To me it feels like there is no way of telling if it is. It will actually be resolved, and I don't know even if it is resolved, if they're closed or what happens and it doesn't feel like it can get very good visibility very easily. So, it just makes it feel like your stuff sits for so long and this is a very good feature, and I really wouldn't like it to sit forever. I did upvote this on discussions.

**Comment by Solarunit** [comment]
I don't think that we will see it near future, even google assistant is not good in that.
IMO the first step should be a very reliable voice detection in bad condidtions like when TV or music is playing in the background.

**Comment by TheBest-Name** [comment]
I get what you're saying, but I do think that's two separate problems that can be worked at the same time.
I think the thing is that this is something that would happen after the audio pipeline has already cleared up the audio. So, it would be behind everything else. Well, or at least it would be in parallel to the actual speech to text and any other things that use the audio for use with controlling things or communicating.
For the actual audio pipeline that you could clear up and do anything you want with the audio before it actually reaches this or the speech to text or the communication or anything else. So, I think it would be OK to add this as long as you just keep it separate, or at least after the audio pipeline has already cleared everything up that way if you ever change the audio pipeline in the future, this will still be OK.

**Comment by TheBest-Name** [comment]
I also want to say just because the audio may not be fully there yet doesn't mean that this is not a good feature to try for, even if the other ones like Google Assistant isn't that good that doesn't mean that this shouldn't eventually be tried for at least.

I will say in the discussion post they do already have some things in there that could be useful. I didn't really look into them, but they did talk about them.

**Comment by Solarunit** [comment]
I mean looks like this feature is too complex that even Google failed to implement it.

**Comment by TheBest-Name** [comment]
Yes, that's true. But I feel like if we keep it modular. Then at least it will be something that will be easier to at least plan for in the future. More just to add on into the side, especially if we used a pre-built solution. And over time, that can get better as well.
I guess what I'm really trying to say is having this wouldn't hurt, all it would do is add extra information, and even if that information is wrong, then there's nothing really lost, since otherwise you wouldn't have had it anyway.

**Comment by TheBest-Name** [comment]
Has this been decided as a no?
It would be a nice feature, but I understand if it is not something that will be implemented.

**Comment by rolyantrauts** [comment]
Alexa and Google do this as you actually enroll a profile as its part of the targetted speech extraction they provide.
If you search github for 'voice fingerprinting' there are many repo's to choose from.
Because I split and have a central authoritive wakeword server on something with more compute it could easily be achieved with https://github.com/rolyantrauts/BoWWClient / https://github.com/rolyantrauts/BoWWServer_x86/tree/main

I just have not implemented it as the audio delivery is open ended to any further ML processes.
I could not tell you how much compute it takes as just have never benchmarked it such as https://github.com/orgs/home-assistant/discussions/527#discussion-8675242 but there are many to choose from.

In fact someone has already done so https://github.com/EuleMitKeule/speaker-recognition
they are using https://github.com/resemble-ai/Resemblyzer
uses the paper https://arxiv.org/pdf/1710.10467
Its not a massive model but brief glance at the paper didn't seem to give much metrics to decide but ask https://github.com/EuleMitKeule what it needs to run.
Its quite a tiny conformer model that you could run https://huggingface.co/tflite-hub/conformer-speaker-encoder and test.
its actually quite easy to detect a speaker but as a secure biometric its very easy to fake.
But for a home system for Bill says 'quarkybeddingspot' the AI could reply 'Bill stop impersonating phrases of Slartibartfast !!!'

**Comment by TheBest-Name** [comment]
Thats cool, we referenced the same discussion thread.

**Comment by TheBest-Name** [comment]
I just want to see who is still watching this and see if you have any updates or things to add.

**Comment by rolyantrauts** [comment]
I just added another option to BcResnet for arcface where rather than standard models it creates embeddings so you can test the distance to recorded wakeword of users.
https://github.com/rolyantrauts/bcresnet/tree/main/models/main-arcface

**Comment by TheBest-Name** [comment]
When you say recorded wake word of users, do you mean wake words that were recorded by the user like samples, or do you mean different wake words that the user can use?

**Comment by rolyantrauts** [comment]
Recorded by the user, you just run through and save the embedding output of those samples and then compare how close those are to incoming wakeword.
Organise them in profiles and you have ID

**Comment by TheBest-Name** [comment]
Very cool. How many profiles can you have and how hard is it to run? Is it better to have multiple recordings for one profile?

**Comment by rolyantrauts** [comment]
Its just compute really and that there is actual difference in the voice profiles, a house / family should be no problem and very low compute. Multiple yeah better, to include distance and emotion styles. 
You can see the embeddings of each class in https://github.com/rolyantrauts/bcresnet/blob/main/models/main-arcface/arcface_golden_weights.json which is the average of each class of the dataset it was trained on and also you can use that.

**Comment by TheBest-Name** [comment]
Just to make sure I know what you're talking about. What do you mean each class? Do you mean like profile or group of recordings.

I think I just found out that it could be what you mean is the wake words.

Also, I'm assuming this, but can you change how much it takes to trigger it? Turning it up to be more likely to trigger down to be less likely.

**Comment by rolyantrauts** [comment]
Classes are what its been trained on 'hey_jarvis', 'noise', 'unknown' wakeword models are classification models.
https://github.com/rolyantrauts/bcresnet/blob/main/models/main-arcface/arcface_golden_weights.json is the average embedding of each class.

**Comment by TheBest-Name** [comment]
When you have those classes and then you take your recordings of people saying those wake words. Is it searching for the wake word and then the recordings or is it trying to filter out the recordings it searches through by the wake world used?

**Comment by rolyantrauts** [comment]
It converts 1.4 sec into a 60 features represented as f32 creating its vector space, using cosine similarity such as arcface you can test by converting other samples to similar 60 feature embeddings and test how close they are together...
Go Google :)
It doesn't filter anything just converts sample audio to embeddings so you can test how close that is to another embedding.
Is this near me saying 'hey jarvis' is done via some fairly simple math of cosine similarity such as arcface by comparing to embedding results, one one saved of you saying 'hey jarvis' and the input.

**Comment by TheBest-Name** [comment]
OK. I think I understood what you're saying. To make sure though, you're saying that the embeddings are the wake words that have been recorded. So, there's not specifically a wake word, just a recording of someone saying a wake word so someone can say any word that they want to wake it up, essentially. Right?

Also, you mentioned Google. Is this the same thing Google uses?

**Comment by rolyantrauts** [comment]
Nope you haven't as it will take any input audio but only spoken wakeword by the same voice will be very close to the embedding of that wakeword spoken by the same voice.

https://arxiv.org/pdf/1801.07698 as in Go Google, go and research yourself... :)

**Comment by TheBest-Name** [comment]
I just glanced over it and I saw that it was for faces, face recognition.

When I said what I said before, I meant that I thought I would take any input audio, but it would match to any audio that was recorded and embedded by the same voice.
So, I think we were thinking the same thing. I was just wondering if that means that you didn't need to have a specific pretrained model of awake word, but instead could just make your own awake words by just saying it and using your voice to have an embedding.

**Comment by rolyantrauts** [comment]
Wake word you convert to an image, so doesn't matter if an audio image, face image or my bum :)
Still does the same and looks for matches.

**Comment by TheBest-Name** [comment]
How do you convert an audio recording of a wake word into a picture? Like a spectrogram?

**Comment by rolyantrauts** [comment]
Spectrogram / logmel / mfcc are images of audio not audio. That is what wakeword models do they convert to images.

You still have to train a specific model but a user doesn't have to retrain but just provide recordings to make it recognise them and be more accurate.
Its like openwakeword but much smaller as rather than create embeddings of all words it just creates embeddings for the classes you provide for that wake word and likely could be more accurate whilst much lighter.

**Comment by TheBest-Name** [comment]
So would you be able to, let's say, have a user add a new wake word by just saying a different word and embedding that.

**Comment by rolyantrauts** [comment]
Nope its for a wakeword currently models provided are 'hey jarvis'

**Comment by TheBest-Name** [comment]
OK, so the recordings that the user records are kind of like extensions of the wake word model that helps it recognize the user speaking.

**Comment by rolyantrauts** [comment]
And wakeword its a different method than softmax or sigmoid activation. Go Google :)

**Comment by TheBest-Name** [comment]
Well, first, how hard do you think it would be to implement it into this Linux Voice Assistant.

I'm not sure if anyone else will see this conversation. I hope they do, because this sounds like exactly the thing this conversation was meant for.

**Comment by rolyantrauts** [comment]
Linux Voice Assistant merely uses the tflite runtime from google as both open and micro wakeword are just refactors and rebrands of existing models and training routines in tensorflow.
https://github.com/arcosoph/nanowakeword or https://github.com/livekit/livekit-wakeword and the above could also be implemented, but usually its ignored unless it has the OHV branding and by OHV devs.

**Comment by TheBest-Name** [comment]
Well, maybe they 'd be willing to refactor or something with this.
I don't know. Maybe we can try to see if they're at least give an opinion on it. Do you think if we mentioned them or something that they would look at it?
I'm just saying that because it seems like this is basically a pre-built solution somewhat that would close this issue. Which I think they would be happy about.

**Comment by rolyantrauts** [comment]
I only create code to share knowledge as not bothered if it is or not.

**Comment by TheBest-Name** [comment]
Is or not what?

**Comment by rolyantrauts** [comment]
Used or ignored.

**Comment by TheBest-Name** [comment]
I see.
Still though, it would be nice to have it just for the features standpoint, even if you don't really care if they use yours in particular.

For me, I would like this feature, which is why I made this request in the first place. So, it would be nice if the feature was added.

**Comment by rolyantrauts** [comment]
Its not just for ID its an alternative wakeword activation method, I just produce proof of concept code just highlight the base bcresnet model is SoTa, opensource and ignored...
There are many methods to do voiceid, differing levels of compute and accuracy, with many repo's and examples.
https://github.com/Qualcomm-AI-research/bcresnet is 3 years old, https://github.com/breizhn/DTLN/tree/master 6 years old, so you might have some waiting to do.

**Comment by TheBest-Name** [comment]
Well, in your opinion, which do you think would be the best for this application?

What I mean is compared to what you've already said and all the other ones that you know of.

**Comment by rolyantrauts** [comment]
There isn't really as its down to the hardware you want to run, as users have a choice of a wide range of hardware, also there needs to be a wide range of models for various tasks.
HA is extremely good as it connects to and brings in all these differing often proprietary, closed source devices to an opensource server.
OHV is almost the opposite where all have to be converted to this strange and over complex OHV protocol, that unless provided by certain contributors/dev its likely to be ignored.
The strangely names voice stack of OHV which currently is unprocessed audio to off the shelf ASR has been that way for 4 years.
Its not software or opensource availability that is the problem.

**Comment by TheBest-Name** [comment]
Yeah, well, I don't think they will change that though. They would have to probably redo the entire thing, and I don't think they're willing to do that. I mean, I guess if it works good enough, then they're going to just take what they already have.

To be fair, it does seem to me, and this could just be my opinion, but that their voice assistant is kind of basic in terms of just taking it in, putting it to an LLM and then spitting out text.

Then again, I suppose that's mostly all it needs to do. But I think that's one of the good things about this repository is maybe this in and of itself can add a more modular feel and add more features to what is already available.

All that being said though, I would imagine people who put this on their device are going to have a very low powered device. I know for me that seems to be the case. I know the other one you had mentioned, arcface. You said that was something that could run on relatively low power compute, if my memory serves me correctly.

**Comment by rolyantrauts** [comment]
You don't need a LLM to turn on a lightbulb...
You can still have NLP processed speech without rigid api's such as [hassil](https://github.com/OHF-Voice/hassil) which aptly named is a lot of hassle just for intents https://github.com/OHF-Voice/intents that 4 years later they are still trying to complete in this strange idea of a multi-language API.
It would be like having a native language version of python for all and the chaos that would cause and why its a bit bat shit crazy.
https://github.com/rolyantrauts/Parakeet2HA using parakeet uses the already existing /api/websocket to get devices and domain and the presentation layer of HA available on the same API gives translations.

They don't have a voice stack as its just unprocessed audio to an ASR, much is refactored and rebranded leaving a knowledge gap on how you could improve and implement in different ways.
There is a lot of delusional aspiration, that has dodged adding all manner of models from fingerprinting, source separation, filters.. 3rd party ready to use opensource, is ignored unless they can claim it is theirs and their has been near zero progress for 4 years.
Some of this is heavy in compute as much is unoptimised, non quantised demonstration PoC models and likely will always be that way. That is why I keep repeating you can have low cost, low compute edge devices of simple function aka Broadcast-On-Wakeword and share a single central unit for the models of higher compute, or if you have cash to burn create standalone devices that for the majority of the time are just idle. 

The devs behind the above have been self congratulatory, 4 years ago i was trying to explain how bad it is https://community.rhasspy.org/t/thoughts-for-the-future-with-homeassistant-rhasspy/4055/3 but I think the lack of contribution from others likely means I am not alone in this. I think I was much stronger in opinion, but think my really salty responses got deleted. In that questioning why ignore a international standard such as websockets, purely to create own-brand, that without collaboration or partners, of a single dev, by sticking a OHV label on it, its instant open standard...
'The Open Home Foundation fights for the fundamental principles of privacy, choice, and sustainability for smart homes. And for every person who lives in one. It does this by supporting the development of open source projects, and open connectivity and communication standards.' 
BS as it totally ignores open source projects, and open connectivity and communication standards unless they are by HA devs that can be branded as OHV.
Any proprietary pythonic diarrhea is posted as a open standard and its just sheer snake oil and because of possible commercial delusions the arena has this garbage trail of these so called opensource next gen voice assistants, that contain many of the same actors, just repeating the same.
The change where HA was this platform to link together all these protocols into a single opensource system, to this land grab of ignoring existing opensource until a HA version can be made has likely also not gone unnoticed.

Some of care about opensource and collaboration and the OSS model is more than just access to readable code.
That OHV is not open and this should not be called Linux Voice Assistant, as its blatant this is being deliberately engineered to be HA only and that is the opposite to why a system like HA garnered success and Linux.
You might believe otherwise but the stagnation of the last 4 years, the fail of VoicePE where the only solution from the devs is to buy-in commercial algs from Xmos,  bundled with a microcontroller not supported by esp-home, that can never be opensource because of xmos licensing...
Worst of all its not even very good.
I am only posting ideas and PoC of ideas I had 4 years ago, through annoyance, that key actors have done so badly it has totally stagnated. I am deliberately posting, to question why so much is ignored and seems to have zero dev. OHV has opened up by adding discussion that we are now engaged in, but for some the last 4 years have likely left a tainted impression.
Some of us are wondering and waiting to see if they can actually do any voice stack dev, than just wrapping python around exiting frameworks and claiming this is a Linux Voice Assistant that is merely ASR / TTS, missing the voice stack that is speech-enhancement and various of speech related ML.

**Comment by TheBest-Name** [comment]
I'm just wondering, maybe they're not sure exactly how to implement some of the stuff you're wanting to implement into the stuff they already have?
What I am getting at is that you seem to know a lot about the audio stack and have written code and trained some of these models it seems. So maybe you could try submitting a pull request.
I say this, but you have to remember that if they want a dev to be doing this and for it to be something that Dev did, if you become the dev, then that would give them what they want and then you could make it more open source by having an audio stack that make sense.

**Comment by rolyantrauts** [comment]
I am not interested in being a Dev, I have just done some PoC examples to demonstrate in a couple of months a voice stack that has been lacking for 4 years.
They do need a dev who understands ML and audio engineering and its obviously lacking, when someone disabled with MS frustrated at the lack of progress can publish so many valid in options in just 2 months.
I used to be quite a good professional dev / computer consultant and merely do a few things, to help with the depression and uselessness of disability, to help push cognitive usage.
I know my limitations and know what should be needed and it isn't me, but it does speak volumes about what is currently being produced or more correctly the lack of.

**Comment by TheBest-Name** [comment]
Well, if you're sure. Still though, with the examples you did provide, maybe they could still use that in a pull request and since most of it has been already looked at, then maybe it would be easier for them to just add in as a module and they might appreciate that.

That way you wouldn't have to be a dev, but the dev would also get some of the work handled.

**Comment by rolyantrauts** [comment]
Its all MIT deliberately licensed and I am hoping the noise being made will actually trigger something than this 4 year stagnation.
Doesn't even require credit and hence why the most liberal / open licensing was used and is licenced.
I am on my longest remission, but this is an exercise for cognitive function, typing co-ordination and just to give me something to do, based on what I did previously professionally.
Which actually was a total different field as I am a RAD / DB guy really, Pi's and voice system just caught my interest 7 years ago and is a hobby.
That is why I find much of what is on offer so shocking as it should be far more advanced than the likes of myself, but the opposite would seem to be true.

**Comment by TheBest-Name** [comment]
I guess it all depends on what you have an interest in. It seems like when you have an interest, it's much easier and quicker to learn things.

Though that being said, you did mention a lot of options, and I do think I would want to have someone else think about whether or not they could implement it in this repository.

Because the point of this issue in the first place was to try to get this feature and you have seemed to have laid out a good way in order to get this, so I would like to try to move it forward in that direction as well.

**Comment by rolyantrauts** [comment]
Do as there are many available models to do this, likely what I have done is likely the lightest method, accuracy I dunno, but its also a way to detect the wakeword over other words.
There are many ways to do this that have been available for many years and my beef is why haven't they...

**Comment by TheBest-Name** [comment]
Sorry, my computer was freaking out over GitHub. So now I'm on my laptop, so sorry I couldn't message you sooner.

You see, I would like this feature and you seem to know the most about how that it actually works behind the scenes in the audio, and you know more about different models at least than me. That's why I would like you to be able to talk to someone who actually is a dev, who would be able to at least look at implementing something, whereas I can only try to know what you're talking about.

Again, this would be the perfect issue to do this in because this is exactly what the issue is trying to get at is to add this feature, so I do hope that someone comes on and looks at adding this. Although I do understand if it doesn't happen, but it would be a nice feature in my opinion at least. And other people have apparently wanted it in the discussion posts that we have referenced.

**Comment by TheBest-Name** [comment]
I just wanted to see what you thought about where we were right now in terms of this discussion.

**Comment by TheBest-Name** [comment]
Okay I just thought I'd check in one more time for now.
